apache/superset
# feat: Create D3NumberFormatters with active Locale

### SUMMARY
Currently, numbers and currencies are formatted using the en Locale, i.e. with group separator `,` and decimal separator `.`

There have been a couple of issues reports raising the need to customize superset and proposing ways to do so (e.g. #https://github.com/apache/superset/issues/3972 and https://github.com/apache/superset/issues/8913).

This (draft) PR suggests to apply the currently active Locale to do number Formatting (though this might not be the best solution, I'm open for feedback...)

### Open questions
While implementing this proposal, I had a number of questions:
1. Should number formats really change when changing the Locale, as just changing the unit without doing currency conversion is plain false (For currencies, adding a special Formatter as proposed in https://stackoverflow.com/questions/63319902/customise-the-number-format-in-apache-superset/63385549#63385549 might be the better way to do so...)
2. Should superset copy D3's locale defintions or just retrieve them from CDN (Note that a Locale conversion from superset to D3 would be necessary, as highlighted by @craigappl [here](https://github.com/apache/superset/issues/3972#issuecomment-447307711)
3. There are a couple of other places where numbers are formatted, which are not addressed by this PR (e.g. table chart)

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Before:
<img width="875" alt="grafik" src="https://user-images.githubusercontent.com/2187389/146511296-cbdff3ab-705e-44da-b493-55bb10f8b343.png">

After:
<img width="869" alt="grafik" src="https://user-images.githubusercontent.com/2187389/146511136-262970da-2d5b-4aca-8e3f-fcef6ca45b9f.png">


### TESTING INSTRUCTIONS
1. add "de": {"flag": "de", "name": "Deutsch"}, to the list of LANGUAGES in superset_config.py
2. start superset an switch to German (Deutsch)
3. display some charts and see ther number format changed.

### ADDITIONAL INFORMATION
- [x] Has associated issue: #3972 #8913
